File: superstitioMain/java/superstitio/powers/SexualDamage.java
package superstitio.powers;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.evacipated.cardcrawl.mod.stslib.powers.interfaces.HealthBarRenderPower;
import com.megacrit.cardcrawl.actions.AbstractGameAction;
import com.megacrit.cardcrawl.cards.DamageInfo;
import com.megacrit.cardcrawl.core.AbstractCreature;
import superstitio.DataManager;
import superstitio.delayHpLose.UnBlockAbleIgnoresTempHPDamage;
import superstitioapi.InBattleDataManager;
import superstitioapi.SuperstitioApiSubscriber;
import superstitioapi.cards.DamageActionMaker;
import superstitioapi.powers.interfaces.OnPostApplyThisPower;

import static superstitioapi.InBattleDataManager.subscribeManageGroups;

public class SexualDamage extends AbstractSuperstitioPower implements HealthBarRenderPower, OnPostApplyThisPower<SexualDamage>,
        SuperstitioApiSubscriber.AtEndOfPlayerTurnPreCardSubscriber {
    public static final String POWER_ID = DataManager.MakeTextID(SexualDamage.class);
    private static final Color BarColor = Color.PURPLE.cpy();
    protected final AbstractCreature giver;

    public SexualDamage(final AbstractCreature owner, int amount, AbstractCreature giver) {
        super(POWER_ID, owner, amount);
        this.giver = giver;
//        ReflectionHacks.setPrivate(this, "greenColor", Color.PINK.cpy());
    }

    @Override
    public void updateDescriptionArgs() {
        setDescriptionArgs(this.amount);
    }

    @Override
    public int getHealthBarAmount() {
        return this.amount;
    }

    @Override
    public Color getColor() {
        return BarColor;
    }

    @Override
    public void renderAmount(SpriteBatch sb, float x, float y, Color c) {
        float temp = this.fontScale;
        this.fontScale *= 1.5f;
        super.renderAmount(sb, x, y, c);
        this.fontScale = temp;
    }

    @Override
    public void receiveAtEndOfPlayerTurnPreCard() {
//        this.owner.damage(BindingHelper.makeInfo(new DamageModContainer(this, new UnBlockAbleDamage()), giver, amount, DamageType.HP_LOSS));
        this.flash();
        DamageActionMaker.maker(this.giver, this.amount, this.owner)
                .setDamageModifier(this, new UnBlockAbleIgnoresTempHPDamage())
                .setDamageType(DamageInfo.DamageType.HP_LOSS)
//                .setDamageType(   DataManager.CanOnlyDamageDamageType.UnBlockAbleDamageType)
                .setEffect(AbstractGameAction.AttackEffect.POISON)
                .addToTop();
        addToBot_removeSpecificPower(this);
    }

    @Override
    public void InitializePostApplyThisPower(SexualDamage addedPower) {
        if (subscribeManageGroups.stream()
                .anyMatch(iSubscriber -> iSubscriber instanceof SexualDamage && ((SexualDamage) iSubscriber).owner == this.owner))
            return;
        InBattleDataManager.Subscribe(this);
    }

    @Override
    public void onRemove() {
        InBattleDataManager.UnSubscribe(this);
    }
}

File: superstitioMain/java/superstitio/powers/SexualDamage.kt
package superstitio.powers

import basemod.interfaces.ISubscriber
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.evacipated.cardcrawl.mod.stslib.powers.interfaces.HealthBarRenderPower
import com.megacrit.cardcrawl.actions.AbstractGameAction.AttackEffect
import com.megacrit.cardcrawl.cards.DamageInfo.DamageType
import com.megacrit.cardcrawl.core.AbstractCreature
import superstitio.DataManager
import superstitio.delayHpLose.UnBlockAbleIgnoresTempHPDamage
import superstitioapi.SuperstitioApiSubscriber.AtEndOfPlayerTurnPreCardSubscriber
import superstitioapi.actions.DamageEnemiesAction
import superstitioapi.powers.interfaces.OnPostApplyThisPower
import superstitioapi.renderManager.inBattleManager.InBattleDataManager
import superstitioapi.utils.PowerUtility
import superstitioapi.utils.addToBot_removeSelf
import superstitioapi.utils.setDescriptionArgs

class SexualDamage //        ReflectionHacks.setPrivate(this, "greenColor", Color.PINK.cpy());
    (owner: AbstractCreature, amount: Int, protected val giver: AbstractCreature) :
    AbstractSuperstitioPower(POWER_ID, owner, amount), HealthBarRenderPower, OnPostApplyThisPower<SexualDamage>,
    AtEndOfPlayerTurnPreCardSubscriber
{
    override fun updateDescriptionArgs()
    {
        setDescriptionArgs(this.amount)
    }

    override fun getHealthBarAmount(): Int
    {
        return this.amount
    }

    override fun getColor(): Color
    {
        return BarColor
    }

    override fun renderAmount(sb: SpriteBatch, x: Float, y: Float, c: Color)
    {
        val temp = this.fontScale
        this.fontScale *= 1.5f
        super.renderAmount(sb, x, y, c)
        this.fontScale = temp
    }

    override fun receiveAtEndOfPlayerTurnPreCard()
    {
//        this.owner.damage(BindingHelper.makeInfo(new DamageModContainer(this, new UnBlockAbleDamage()), giver, amount, DamageType.HP_LOSS));
        this.flash()
        DamageEnemiesAction.builder(this.amount, this.owner)
            .setSource(this.giver)
            .setDamageModifier(this, UnBlockAbleIgnoresTempHPDamage())
            .setDamageType(DamageType.HP_LOSS)
            .setAttackEffectType(AttackEffect.POISON)
            .setAfterDamageConsumer { PowerUtility.RemovePower(this.owner, this) }
            .addToTop()
        addToBot_removeSelf()
    }

    override fun InitializePostApplyThisPower(addedPower: SexualDamage)
    {
        if (InBattleDataManager.subscribeManageGroups.stream()
                .anyMatch { iSubscriber: ISubscriber? -> iSubscriber is SexualDamage && iSubscriber.owner === this.owner }
        ) return
        InBattleDataManager.Subscribe(this)
    }

    override fun onRemove()
    {
        InBattleDataManager.UnSubscribe(this)
    }

    companion object
    {
        val POWER_ID: String = DataManager.MakeTextID(SexualDamage::class.java)
        private val BarColor: Color = Color.PURPLE.cpy()
    }
}
